Signal/src/ViewControllers/ConversationView/Cells/OWSContactShareButtonsView.m

```diff
@@ -0,0 +1,161 @@
+
+@property (nonatomic, readonly) ContactShareViewModel *contactShare;
+@property (nonatomic, weak) id<OWSContactShareButtonsViewDelegate> delegate;
+
+@property (nonatomic, readonly) OWSContactsManager *contactsManager;
+
+@property (nonatomic, nullable) UIView *buttonView;
+
+@end
+
+#pragma mark -
+
+@implementation OWSContactShareButtonsView
+
+- (instancetype)initWithContactShare:(ContactShareViewModel *)contactShare
+                            delegate:(id<OWSContactShareButtonsViewDelegate>)delegate
+{
+    self = [super init];
+
+    if (self) {
+        _delegate = delegate;
+        _contactShare = contactShare;
+        _contactsManager = [Environment current].contactsManager;
+    }
+
+    return self;
+}
+
+#pragma mark -
+
++ (BOOL)hasSendTextButton:(ContactShareViewModel *)contactShare contactsManager:(OWSContactsManager *)contactsManager
+{
+    OWSAssert(contactShare);
+    OWSAssert(contactsManager);
+
+    return [contactShare systemContactsWithSignalAccountPhoneNumbers:contactsManager].count > 0;
+}
+
++ (BOOL)hasInviteButton:(ContactShareViewModel *)contactShare contactsManager:(OWSContactsManager *)contactsManager
+{
+    OWSAssert(contactShare);
+    OWSAssert(contactsManager);
+
+    return [contactShare systemContactPhoneNumbers:contactsManager].count > 0;
+}
+
++ (BOOL)hasAddToContactsButton:(ContactShareViewModel *)contactShare
+{
+    OWSAssert(contactShare);
+
+    return [contactShare e164PhoneNumbers].count > 0;
+}
+
++ (BOOL)hasAnyButton:(ContactShareViewModel *)contactShare
+{
+    OWSAssert(contactShare);
+
+    OWSContactsManager *contactsManager = [Environment current].contactsManager;
+
+    return [self hasAnyButton:contactShare contactsManager:contactsManager];
+}
+
++ (BOOL)hasAnyButton:(ContactShareViewModel *)contactShare contactsManager:(OWSContactsManager *)contactsManager
+{
+    OWSAssert(contactShare);
+
+    return ([self hasSendTextButton:contactShare contactsManager:contactsManager] ||
+        [self hasInviteButton:contactShare contactsManager:contactsManager] ||
+        [self hasAddToContactsButton:contactShare]);
+}
+
++ (CGFloat)bubbleHeight
+{
+    return self.buttonHeight;
+}
+
++ (CGFloat)buttonHeight
+{
+    return MAX(44.f, self.buttonFont.lineHeight + self.buttonVMargin * 2);
+}
+
++ (UIFont *)buttonFont
+{
+    return [UIFont ows_dynamicTypeBodyFont].ows_mediumWeight;
+}
+
++ (CGFloat)buttonVMargin
+{
+    return 5;
+}
+
+- (void)createContents
+{
+    OWSAssert([OWSContactShareButtonsView hasAnyButton:self.contactShare contactsManager:self.contactsManager]);
+
+    self.layoutMargins = UIEdgeInsetsZero;
+    self.backgroundColor = [UIColor ows_light02Color];
+
+    UILabel *label = [UILabel new];
+    self.buttonView = label;
+    if ([OWSContactShareButtonsView hasSendTextButton:self.contactShare contactsManager:self.contactsManager]) {
+        label.text = NSLocalizedString(@"ACTION_SEND_MESSAGE", @"Label for 'sent message' button in contact view.");
```

Comment:
comment typo: "send message" also, I don't think the context of "contact view" is sufficiently clear.  e.g. the context in line 124 is better.

Reply:
Done.